### DISCUSSION TITLE
When is StreetComplete uploading changes to OSM?

### DISCUSSION THREAD

**Discussion by simonschaufi** [streetcomplete/StreetComplete#5116] [Q&A] [answered]
Hi, I would like to know what triggers the upload of changes to OSM as there is no "upload" button. People started complaining about massive changesets with changes from all around around the world because I couldn't split them up or I just don't know HOW to split them up. See https://www.openstreetmap.org/changeset/137939866.

Can you please explain how the sync works in detail?

**Comment by mnalis** [comment, marked as answer]
To the best of my understanding (but I am not authorative source): 

Depending on `Settings` / `Upload answers automatically`:

- if `Off`, there is manual upload arrow on the main screen, and only when it is pressed, _upload starts_
- if `On`, _upload starts_ immediately after solving each quest
- if `Only on Wi-Fi`, _upload starts_ only when phone connects to Wi-Fi network (it also requires StreetComplete to be active app in the foreground)

Take note that _"upload starts"_ means:
- if there is no working Internet connection, nothing can be uploaded (obviously)
- otherwise, all yet unuploaded quest answers are uploaded in an already open changeset (or a new changeset is created if there is no active one for that quest type)
- also note that each quest type has its own changeset with different comment (e.g. there is one active changeset for `Survey whether payment with cards is accepted`, one for `Specify paths smoothness`, one for `Specify whether there are cycleways` etc. - depending what quests you've been solving).

The changesets are closed automatically on timeout (I don't think SC ever closes them manually). According to the wiki https://wiki.openstreetmap.org/wiki/Changeset#Opening_and_closing_changesets that happens when any of those conditions are reached:

- if no uploads happen in that changeset for 1 hour (idle timeout)
- if changeset has been open for 1 day (maximum open time)
- if changeset has more than 10,000 elements (maximum capacity) - unlikely to happen with StreetComplete unless you're on a rampage :smile_cat:

**Reply by mnalis** [reply]
That being said, it would be interesting to hear what you did exactly in that changeset? Travel from Germany to New Zealand? It might be that there was no active internet connection when you edited one country before the trip, and then you edited another country, and then you got internet - and thus they got uploaded simultaneously in same changeset. If so, the solution is to make sure to upload answers for one country (i.e. have working internet connection if you're using auto-upload) at least 1h before  traveling to other country.

But that should happen quite rarely. Are there more changesets like that where people complained (as it seems implied in your question?), or just that one? 

If it happens much more often, please note that StreetComplete is intended to be used only for surveying on-the-ground (and not as a general purpose editor like iD or JOSM) - is that how you're using it?

**Reply by simonschaufi** [reply]
I wasn't travelling. I'm currently in New Zealand but lived before in Germany and downloaded a small area there and made a few changes that I could do out of my head knowledge but I didn't think that it would become such a massive changeset. I did all changes within a relatively short time and had the settings to "auto".

I know for example that Organic Maps does the upload when the App goes into background and I think I also did that a couple of times but if StreetComplete works different, then it should be at least explained somewhere what "auto" mode really does.

> But that should happen quite rarely. Are there more changesets like that where people complained (as it seems implied in your question?), or just that one?

It was just one and looking at the history, I saw that SC split the changes "somehow" up into different categories. That is why I thought that it would also do that for a specific country to not have these massive changesets but I was wrong.

Compare:
- https://www.openstreetmap.org/changeset/137940180 (Specify building types) with
- https://www.openstreetmap.org/changeset/137939866 (Specify road surfaces) and
- https://www.openstreetmap.org/changeset/137939993 (Survey opening hours)

The other 2 are very small but of the same category.

> If it happens much more often, please note that StreetComplete is intended to be used only for surveying on-the-ground (and not as a general purpose editor like iD or JOSM) - is that how you're using it?

I was changing street surfaces and even without being on that road, I know that all roads within Cambridge are of the same material so that was a rather easy "fix".
